3.13/specializer.md

```diff
@@ -0,0 +1,289 @@
+following Python code:
+
+```python
+x = a + b + c
+call(thing)
+```
+
+It will produce the following IR (assuming no guards are required):
+```
+x    := BINARY_OP_ADD_INT(a1, BINARY_OP_ADD_INT(b1, c1))
+null := CALL(thing1)
+```
+
+When guards are required, these are raw instructione entries:
+```
+stack0  := INIT_FAST a1
+stack1  := INIT_FAST b1
+ir_inst(GUARD_BOTH_INT)
+x       := BINARY_OP_ADD_INT(c1, BINARY_OP_ADD_INT(stack0, stack1))
+null    := CALL(thing1)
+```
+
+The author previously experimented with SSA graphs (Lemerre, 2023)
+for the IR, and researched the Sea of Nodes (Click & Paleczny
+, 1995). However, we have decided with this simpler IR for a few
+reasons:
+
+1. It is requires less memory to represent.
+2. It is less complex.
+3. Guards represent points where CPython state has to be consistent
+between tiers for cheap and easy deoptimization. The graph-based IRs
+currently do not guarantee that without some form of ordering and state
+information encoded in. A graph-based IR may be more suitable for Tier 3,
+where reconstructing the stack and locals state of the interpreter may
+be feasible.
+4. Guards also represent strict dependencies in control-flow, requiring
+more instruction scheduling in graph-based IRs.
+
+By using a hybrid of symbolic terms, assignments, and plain instructions,
+we inherently encode the proper order of instructions, while obtaining
+some benefit from graph-based IRs: namely their ability to combine
+multiple optimizations into a single pass (which you will see below).
+
+## Optimization
+
+All the following optimizations are done in a single pass together with
+the analysis stage:
+
+### Type Propagation
+We perform these automatically with the abstract interpreter's symbolic
+terms. As part of a term's type, we also encode auxillary type information
+in the type system, for example, dictionary versions or type verions.
+
+These rules follow deductive steps. Type information is propagated forwards.
+For example, in the IR:
+```
+stack0  := INIT_FAST a1
+stack1  := INIT_FAST b1
+# types: (a1: unknown, b1: unknown)
+ir_inst(GUARD_BOTH_INT)
+# types: (a1: int, b1: int)
+x       := BINARY_OP_ADD_INT(c1, BINARY_OP_ADD_INT(stack0, stack1))
+# types: (a1: int, b1: int, x: int)
+null    := CALL(thing1)
+```
+
+The key part is that type and constant propagation is done automatically,
+without having to handwrite these deductive rules ourselves. Types are
+expressed as part of the interpreter DSL:
+
+```
+// Typed outputs means the types are these types after the operation.
+pure op(_BINARY_OP_ADD_INT, (left, right -- res: ~(PYINT_TYPE))) {
+}
+
+// Typed output with auxillary type informatino (type_version)
+guard op(_GUARD_TYPE_VERSION, (type_version/2, owner -- owner: ~(GUARD_TYPE_VERSION_TYPE + type_version))) {
+}
+```
+
+The abstract interpreter's then automatically generates all of the
+cases. With types, we can then eliminate type guards that are not
```

Comment:
How does it automatically generate all the cases?
How does it know how to handle `_GUARD_DORV_VALUES_INST_ATTR_FROM_DICT` or `_GUARD_KEYS_VERSION`?

Reply:
DSL:
```
        guard op(_GUARD_KEYS_VERSION, (keys_version/2, owner -- owner: ~(GUARD_KEYS_VERSION_TYPE + keys_version))) {
            PyTypeObject *owner_cls = Py_TYPE(owner);
            PyHeapTypeObject *owner_heap_type = (PyHeapTypeObject *)owner_cls;
            DEOPT_IF(owner_heap_type->ht_cached_keys->dk_version != keys_version);
        }
```
abstract interpreter cases:
```
        case _GUARD_KEYS_VERSION: {
            _Py_UOpsSymbolicExpression *__owner_;
            __owner_ = stack_pointer[-1];
            uint32_t keys_version = (uint32_t)CURRENT_OPERAND();
            // Constant evaluation 
            if (is_const(__owner_)) {
                PyObject *owner;
                owner = get_const(__owner_);
                PyTypeObject *owner_cls = Py_TYPE(owner);
                PyHeapTypeObject *owner_heap_type = (PyHeapTypeObject *)owner_cls;
                if (owner_heap_type->ht_cached_keys->dk_version != keys_version) goto error;

                DPRINTF(3, "const eliminated guard\n");
                break;
            }
            // Type guard elimination 
            if (sym_matches_type((_Py_UOpsSymbolicExpression *)__owner_, GUARD_KEYS_VERSION_TYPE, (uint32_t)keys_version)){
                DPRINTF(2, "type propagation eliminated guard\n");
                break;
            }
            else {
                // Type propagation 
                sym_set_type((_Py_UOpsSymbolicExpression *)__owner_, GUARD_KEYS_VERSION_TYPE, (uint32_t)keys_version);
                goto guard_required;
            }
            break;
        }
```